tools/underwater-gps.py

```diff
@@ -184,79 +276,86 @@ def processLocatorPosition(response, *args, **kwargs):
-s = grequests.Session()
-# Thank you https://stackoverflow.com/questions/16015749/in-what-way-is-grequests-asynchronous
-while True:
-    if time.time() > last_locator_update + update_period:
-        last_locator_update = time.time()
-        url = gpsUrl + "/api/v1/position/global"
-        print('requesting data from', url)
-        request = grequests.get(url, session=s, hooks={'response': processLocatorPosition})
-        grequests.send(request)
-
-    if time.time() > last_master_update + update_period:
-        last_master_update = time.time()
-        url = gpsUrl + "/api/v1/position/master"
-        print('requesting data from', url)
-        request = grequests.get(url, session=s, hooks={'response': processMasterPosition})
-        grequests.send(request)
-
-    try:
-        datagram = sockit.recvfrom(4096)
-        recv_payload = json.loads(datagram[0])
-        if 'depth' not in recv_payload or 'temp' not in recv_payload:
-            print('not valid payload:')
-            print(json.dumps(recv_payload, indent=4, sort_keys=True))
+    result = json.dumps(result)
+    socket_mavproxy.sendto(result, ('0.0.0.0', 25100))
+
+
+# Socket to send GPS data to mavproxy
+socket_mavproxy = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+socket_mavproxy.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+socket_mavproxy.setblocking(0)
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Driver for the Water Linked Underwater GPS system.")
+    parser.add_argument('--ip', action="store", type=str, default="demo.waterlinked.com", help="remote ip to query on.")
+    parser.add_argument('--port', action="store", type=str, default="80", help="remote port to query on.")
+    args = parser.parse_args()
+
+    # Use UDP port 14401 to send NMEA data to QGC for topside location
+    qgcNmeaSocket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+    qgcNmeaSocket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+    qgcNmeaSocket.setblocking(0)
+
+    gpsUrl = "http://" + args.ip + ":" + args.port
+
+    setup_streamrates()
+
+    wait_for_waterlinked()
+
+    report_status("Running")
+
+    # Sets GPS type to MAVLINK
+    set_param("GPS_TYPE", "MAV_PARAM_TYPE_UINT8", 14)
+
+    # update at 5Hz
+    update_period = 0.25
+    last_master_update = 0
+    last_locator_update = 0
+    last_position_update = 0
+
+    depth_endpoint = gpsUrl + "/api/v1/external/depth"
+    ext_depth = {}
+    orientation_endpoint = gpsUrl + "/api/v1/external/orientation"
+    ext_orientation = {}
+    locator_endpoint = gpsUrl + "/api/v1/position/global"
+    master_endpoint = gpsUrl + "/api/v1/position/master"
+
+    # TODO: upgrade this to async once we have Python >= 3.6
+    while True:
+        time.sleep(0.02)
+        if time.time() > last_locator_update + update_period:
+            last_locator_update = time.time()
+            try:
+                response = urllib2.urlopen(locator_endpoint, timeout=1).read()
```

Comment:
I'm seeing a bunch of this try except for urlopen that only report status, it's maybe better to move it to a function